Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  AppleSingInProvider.swift
+//  Core
+//
+//  Created by Eugene Yatsenko on 10.10.2023.
+//
+
+import Foundation
+import AuthenticationServices
+import Swinject
+
+public struct AppleCredentials: Codable {
+    public var name: String
+    public var email: String
+    public var birthYear: String? = nil
+    public var token: String
+}
+
+public final class AppleSingInProvider: NSObject, ASAuthorizationControllerDelegate {
+
+    public override init() {}
+
+    private var completion: ((Result<AppleCredentials, Error>) -> Void)?
+    private let appleIDProvider = ASAuthorizationAppleIDProvider()
+
+    public func request(completion: ((Result<AppleCredentials, Error>) -> Void)?) {
+        let request = appleIDProvider.createRequest()
+        request.requestedScopes = [.fullName, .email]
+
+        let authorizationController = ASAuthorizationController(authorizationRequests: [request])
+        authorizationController.delegate = self
+        authorizationController.performRequests()
+
+        self.completion = completion
+    }
+
+    public func authorizationController(
+        controller: ASAuthorizationController,
+        didCompleteWithAuthorization authorization: ASAuthorization
+    ) {
+        guard let credentials = authorization.credential as? ASAuthorizationAppleIDCredential else {
+            completion?(.failure(CustomError.unknownError))
+            return
+        }
+
+        var storage = Container.shared.resolve(CoreStorage.self)
```

Comment:
It should be injected into the initializer.
Entity shouldn't manage dependency and call DI itself.